Fix NULL pointer dereference that occurs when byte swapping is needed

phonohawk opened the pull request:
cdrom_cache_handler() calls cdio_cddap_read_timed() with buffer == NULL. When this happens on a platform where byte swapping is needed, the latter crashes for dereferencing a NULL pointer.

rocky commented:
Thanks for the patch.

It looks to me like the NULL buffer test is in a block that is a little bit too all-encompassing.

It looks to me that a non-harmful bigendian test is skipped. This is not helpful I guess if the platform _is_ bigendian. And if it isn't, then with a more judicious patch it would be helpful in little-endian platforms as well.

So shouldn't the dereference test go down a little further in:

```
	if ( d->b_swap_bytes && d->bigendianp != bigendianp() ) {
```

?

This code is pretty much the same as in CD-Paranoia. I wonder if they suffer the same problem, or if there's something libcdio-paranoia does that calls with a NULL buffer while the original CD-Paranoia doesn't. Do you have any insight on that?

Should it the case that CD-Paranoia has the same problem you may also want to suggest a patch there as well.

Finally, if there's a way to create a test case to catch this bug, I'd appreciate it if you'd write one.

Thanks, again.

phonohawk commented:
Thanks for your helpful feedback. I updated my commit.

> So shouldn't the dereference test go down a little further in:

Yes, indeed. I overlooked the call of `data_bigendianp(d)` which clearly shouldn't be skipped.

> This code is pretty much the same as in CD-Paranoia. I wonder if they suffer the same problem, or if there's something libcdio-paranoia does that calls with a NULL buffer while the original CD-Paranoia doesn't. Do you have any insight on that?

The original CD-Paranoia, at least its SVN trunk, already has a fix for this: https://trac.xiph.org/changeset/15356/

> Finally, if there's a way to create a test case to catch this bug, I'd appreciate it if you'd write one.

Added a round-trip test from little- to big- to little-endian. It should catch the bug regardless of the host endianness by trying to read BIN/CUE files both in little- and big-endian.

phonohawk commented:
> Does this work for you? It doesn't for me. On my linux box, -c '%s' is used with stat to get file sizes.

Gee... I didn't know that BSD `stat(1)` was incompatible with GNU coreutils. I tested it on Darwin, NetBSD, and FreeBSD and they all worked. I'm looking for a possible workaround...

> The other weird thing that was happening is that even though "stat" gave an error the test succeeded.

```
test `fofdsas` -ne `dfdas`
```
evaluates to `test -ne` in that case, which is obviously an illegal expression. The `test(1)` utility silently succeeds on a syntax error for some reason.

phonohawk commented:
Updated the commit again. It turned out that `stat(1)` was not in the POSIX standard so I replaced it with `wc -c < FILE`.

rocky commented:
Thanks again for putting up with me. I appreciate your attention to detail.
